### PR TITLE
pytorch lambdify nn_module option (produces an actual torch nn.module)

### PR DESCRIPTION
**Pull request by clayton-r**
Does what @MilesCranmer wanted in #27760 by adding a separate mechanism to export as an actual pytorch module rather than using the printer mechanism. 

Why do I think we need both? I think it's worth keeping the the option to have the the printer version as-is while also supporting pytorch module conversion. 

* printing
  * Added new module option to the pytorch printer
* utilities
  * Added PyTorch as module support to lambdify

### CODE REVIEW

File: sympy/utilities/lambdify.py [hunk 1]
```diff
@@ -815,6 +819,10 @@ def _lambdifygenerated(x):
         for term in syms:
             namespace.update({str(term): term})
 
+    if _module_present('torch_module', namespaces):
```
Comment (asmeurer) [inline sympy/utilities/lambdify.py:822]: We need to document this.

File: sympy/printing/pytorch.py [hunk 2]
```diff
@@ -292,6 +471,175 @@ def _print_polygamma(self, expr):
     _ones = "ones"
     _zeros = "zeros"
 
+
+class _Node(torch.nn.Module):
```
Comment (asmeurer) [inline sympy/printing/pytorch.py:475]: The module needs to be importable when pytorch isn't installed.

**Comment by asmeurer** [comment]
Is it possible to make this work by passing in a different printer to lambdify? That might be better than the special "torch_module" module (especially since it's confusing because "module" has dual meanings in that particular context).

**Comment by clayton-r** [comment]
> Is it possible to make this work by passing in a different printer to lambdify? That might be better than the special "torch_module" module (especially since it's confusing because "module" has dual meanings in that particular context).

The current TorchPrinter outputs a lambda that chains pytorch functions (torch.sin, torch.add), with requires_grad support for gradients. It maps to pytorch's functional API and works well in that context.

pytorch's nn.Module system is the actual core of the neural network architecture and includes things like parameters, buffers, etc. under the hood, and these modules are composed with other modules to form the computational graph of the model. 

Rather than going through the usual Printing / string creation mechanism, This PR's SympyModule (via torch_module) builds an nn.Module directly from a sympy expression. It creates a computational graph with _Node submodules. This allows users to integrate the expression into pytorch models, train it, or compose it with other layers—functions the existing pytorch lambda approach cannot handle natively since it won’t have the parameters, etc.

Adding this as a module option maintains the functional approach for lightweight use while providing an nn.Module for pytorch's class-based workflows.

IMO, the existing pytorch lambda version should be kept because it aligns with lambdify's behavior across backends (lightweight callable function) and avoids nn.Module overhead when not needed. We could go with torch_nn rather than torch_module since that’s a common convention in pytorch world.

**Comment by asmeurer** [comment]
I'm on board with the differences between pytorch functions and pytorch modules and the need for this. I'm just concerned with the user API proposed here, which is `lambdify(x, f(x), "pytorch_module")`. I find that to be confusing, and I'm wondering if we could instead do `lambdify(x, f(x), "pytorch", printer=PytorchModulePrinter)`. It's just a end-user API question.

Also, I just noticed that in either case, this is bypassing all the stuff in lambdify that processes the text from the printer. This is to be expected, but some of the things that are skipped could be surprising. Basically, all the functionality from [this line](https://github.com/sympy/sympy/pull/27887/files#diff-22b8dac2993b5b903ba7dd511636fa0c7b753664e59b6da84c8207dd18c0010fR858) down will not be available for pytorch_module. In particular, the cse flag won't work, passing multiple modules won't work, and adding a docstring won't work (and possibly some other things). 

In effect, it isn't really using much of the functionality of lambdify at all. We could try to implement those things, although I wonder if we should just move this to a separate function. The `llvm_callable` function actually works somewhat similarly to this. It doesn't return a string representation, but just returns a direct Python object. I wonder if it would make more sense to make this work like that instead. 

@bjodah do you have any thoughts on this PR?

**Comment by bjodah** [comment]
Honestly, I don't have much experience with pytorch. I think you two both are better qualified to determine the best path forward!

**Comment by asmeurer** [comment]
Keeping the mapping in place is definitely a good idea. We could reuse the printer code. The main thing I'm just not sure about is if it makes sense for lambdify to use something that isn't a printer internally. Then again, maybe it's fine. The approach here does work, it just has some aspects which could be surprising to some users.

**Comment by clayton-r** [comment]
I reworked this with a slightly updated naming convention. Just as a refresh:

Printer (functional API):
  - `lambdify(..., "torch")` stays the functional backend
  - `torch_code(...)` stays the printer path and returns a **string**
  
 Module (pytorch's module/class-based API):
  - `torch_nn_module(...)` returns an actual **`torch.nn.Module` object** like the llvm callable. 
  - `torch_module(...)` is kept only as a backward-compatible wrapper to `torch_nn_module(...)`

 Hopefully the "nn_module" naming convention avoids the “module means two things” confusion in `lambdify(modules=...)` and keeps lambdify behavior predictable.

 I also included the torch scalar/tensor fix for `Max/Min` in lambdify (e.g. `Max(1.0, x)` with tensor inputs), which addresses #28457 / overlaps with #28458.

Since this is conceptually similar to `llvm_callable` (object-returning API) we could put the nn_module stuff there, but this PR keeps printer + module APIs together in `sympy.printing.pytorch` for mapping reuse/maintainability. We can split the module builder later if you prefer that layout.

**Comment by moorepants** [comment]
I just tried this out with an example from SymPy mechanics and it works.

```
conda create -c conda-forge -n sympy-torch pytorch-cpu mpmath ipython
```

Then cd `/path/to/sympy` and checkout this PR.

```
$ activate sympy-torch
```

```python
In [1]: from sympy.physics.mechanics.models import n_link_pendulum_on_cart

In [3]: kane = n_link_pendulum_on_cart(4, False, False)

In [8]: import sympy.physics.mechanics as me

In [9]: x = kane.q[:] + kane.u[:]

In [10]: x
Out[10]: [q0(t), q1(t), q2(t), q3(t), q4(t), u0(t), u1(t), u2(t), u3(t), u4(t)]

In [11]: c = kane.mass_matrix.free_symbols | kane.forcing.free_symbols

In [13]: c.remove(me.dynamicsymbols._t)

In [14]: c
Out[14]: {g, l0, l1, l2, l3, m0, m1, m2, m3, m4}

In [17]: import sympy as sm

In [19]: f = sm.lambdify(x + list(c), (kane.mass_matrix, kane.forcing), modules='torch')

In [20]: import numpy as np

In [22]: import torch

In [23]: x_vals = torch.tensor(np.random.random(20), dtype=torch.float64)

In [25]: f(*x_vals)
Out[25]: 
(tensor([[ 2.2287, -0.6147, -1.0709, -0.0900, -0.0344],
         [-0.6147,  0.2663,  0.4626,  0.0390,  0.0154],
         [-1.0709,  0.4626,  1.0702,  0.0900,  0.0345],
         [-0.0900,  0.0390,  0.0900,  0.0086,  0.0034],
         [-0.0344,  0.0154,  0.0345,  0.0034,  0.0023]], dtype=torch.float64),
 tensor([[-0.0159],
         [ 0.0370],
         [ 0.0200],
         [ 0.0053],
         [ 0.0177]], dtype=torch.float64))

In [26]: %timeit f(*x_vals)
969 μs ± 8.94 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

In [27]: f_numpy = sm.lambdify(x + list(c), (kane.mass_matrix, kane.forcing), cse=True)

In [28]: x_vals_numpy = np.random.random(20)

In [30]: %timeit f_numpy(*x_vals_numpy)
15.3 μs ± 273 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)

In [31]: x_vals = torch.tensor(x_vals_numpy, dtype=torch.float64)

In [32]: f(*x_vals)
Out[32]: 
(tensor([[ 3.0681, -1.4585, -0.5699, -0.3140, -0.5238],
         [-1.4585,  0.8220,  0.3257,  0.2043,  0.2826],
         [-0.5699,  0.3257,  0.1863,  0.1180,  0.1608],
         [-0.3140,  0.2043,  0.1180,  0.1306,  0.1572],
         [-0.5238,  0.2826,  0.1608,  0.1572,  0.2978]], dtype=torch.float64),
 tensor([[-0.3651],
         [ 0.2536],
         [ 0.1021],
         [ 0.0541],
         [ 0.0954]], dtype=torch.float64))

In [33]: f_numpy(*x_vals_numpy)
Out[33]: 
(array([[ 3.06814787, -1.45850661, -0.5698963 , -0.31398073, -0.52381902],
        [-1.45850661,  0.8220251 ,  0.3257107 ,  0.20429148,  0.28262452],
        [-0.5698963 ,  0.3257107 ,  0.18627019,  0.11800163,  0.16084099],
        [-0.31398073,  0.20429148,  0.11800163,  0.13058122,  0.15718089],
        [-0.52381902,  0.28262452,  0.16084099,  0.15718089,  0.29782424]]),
 array([[-0.36508718],
        [ 0.25362824],
        [ 0.10211544],
        [ 0.05412338],
        [ 0.09539067]]))
```

**Comment by clayton-r** [comment]
@moorepants That functionality should have already been there. Here's an example of the Module workflow (this is a Black-Scholes calibration example), which uses  `torch_nn_module` instead of the printer with modules=torch.

```bash
conda create -c conda-forge -n sympy-torch pytorch-cpu mpmath ipython
# then checkout this PR branch in your sympy repo
conda activate sympy-torch
```

```python
In [1]: import sympy as sm
In [2]: import torch
In [3]: from sympy.printing.pytorch import torch_nn_module

In [4]: S, K, r, sigma, T = sm.symbols('S K r sigma T', positive=True)
In [5]: sqrt2 = sm.sqrt(2)
In [6]: d1 = (sm.log(S/K) + (r + sigma**2/2)*T) / (sigma*sm.sqrt(T))
In [7]: d2 = d1 - sigma*sm.sqrt(T)
In [8]: N = lambda z: sm.Rational(1, 2) * (1 + sm.erf(z/sqrt2))
In [9]: call = S*N(d1) - K*sm.exp(-r*T)*N(d2)
In [10]: bs = torch_nn_module((S, K, r, sigma, T), call)
In [11]: bs

Out[11]: SymPyTorchModule(expressions=(-K*(erf(sqrt(2)*(-sqrt(T)*sigma + (T*(r + sigma**2/2) + log(S/K))/(sqrt(T)*sigma))/2)/2 + 1/2)*exp(-T*r) + S*(erf(sqrt(2)*(T*(r + sigma**2/2) + log(S/K))/(2*sqrt(T)*sigma))/2 + 1/2),))

In [12]: S_mkt = torch.tensor(100.0, dtype=torch.float64)
In [13]: K_mkt = torch.tensor([80., 90., 100., 110., 120.], dtype=torch.float64)
In [14]: T_mkt = torch.tensor([0.25, 0.5, 1.0, 1.5, 2.0], dtype=torch.float64)
In [15]: r_mkt = torch.tensor(0.02, dtype=torch.float64)
In [16]: with torch.no_grad():
            market_prices = bs(S_mkt, K_mkt, r_mkt, torch.tensor(0.20, dtype=torch.float64), T_mkt)
            
In [17]: market_prices

Out[17]: tensor([20.4335, 12.4539,  8.9160,  7.0562,  5.9469], dtype=torch.float64)

In [18]: %timeit bs(S_mkt, K_mkt, r_mkt, torch.tensor(0.20, dtype=torch.float64), T_mkt)

81.7 us +- 2.0 us per loop (mean +- std. dev. of 7 runs, 10,000 loops each)

In [19]: class BSCalibrator(torch.nn.Module):
            def __init__(self, bs_module, init_sigma=0.50):
                super().__init__()
                self.bs = bs_module
                self.log_sigma = torch.nn.Parameter(torch.log(torch.tensor(init_sigma, dtype=torch.float64)))

            def forward(self, S, K, r, T):
                sigma_ = torch.exp(self.log_sigma)
                return self.bs(S, K, r, sigma_, T)
                
In [20]: model = BSCalibrator(bs, init_sigma=0.50)

In [21]: opt = torch.optim.Adam(model.parameters(), lr=5e-2)

In [22]: for _ in range(300):
            opt.zero_grad()
            pred = model(S_mkt, K_mkt, r_mkt, T_mkt)
            loss = torch.mean((pred - market_prices)**2)
            loss.backward()
            opt.step()

In [23]: torch.exp(model.log_sigma).item()

Out[23]: 0.19999995587485547

In [24]: loss.item()

Out[24]: 2.7803595407934037e-12

In [25]: with torch.no_grad():
            fitted = model(S_mkt, K_mkt, r_mkt, T_mkt)

In [26]: fitted

Out[26]: tensor([20.4335, 12.4539,  8.9160,  7.0562,  5.9469], dtype=torch.float64)
```

The goal is to define symbolic expression in SymPy, compile to a `torch.nn.Module`, and calibrate parameters via autograd/optimizers.

**Comment by moorepants** [comment]
Thanks. I mistook this as adding torch support entirely. I'm not familiar with what you are doing in your example, so I don't think I have much to offer in a review here. Sorry for the noise.

**Comment by moorepants** [comment]
Would you mind changing the title of the PR to better reflect what you are adding?

**Comment by moorepants** [comment]
I noticed this PR, which may be related: https://github.com/sympy/sympy/pull/14689

File: sympy/utilities/lambdify.py (hunk 1 above)
Comment (clayton-r) [reply sympy/utilities/lambdify.py]: in docs now

**Comment by clayton-r** [comment]
> Would you mind changing the title of the PR to better reflect what you are adding?

done

> I noticed this PR, which may be related: https://github.com/sympy/sympy/pull/14689

yep but there are several other options for exporting to other ml frameworks like Tensorflow, etc. already anyway. I added the functional API in #27760
